lib/jxl/encode_internal.h

```diff
@@ -161,38 +168,157 @@ struct JxlEncoderQueuedInput {
+                      bool force_large_box, uint8_t* output);
+
 // Appends a JXL container box header with given type, size, and unbounded
 // properties to output.
 template <typename T>
 void AppendBoxHeader(const jxl::BoxType& type, size_t size, bool unbounded,
                      T* output) {
-  uint64_t box_size = 0;
-  bool large_size = false;
-  if (!unbounded) {
-    box_size = size + 8;
-    if (box_size >= 0x100000000ull) {
-      large_size = true;
-    }
+  size_t current_size = output->size();
+  output->resize(current_size + kLargeBoxHeaderSize);
+  size_t header_size =
+      WriteBoxHeader(type, size, unbounded, /*force_large_box=*/false,
+                     output->data() + current_size);
+  output->resize(current_size + header_size);
+}
+
+}  // namespace jxl
+
+class JxlOutputProcessorBuffer;
+
+class JxlEncoderOutputProcessorWrapper {
+  friend class JxlOutputProcessorBuffer;
+
+ public:
+  JxlEncoderOutputProcessorWrapper() = default;
+  explicit JxlEncoderOutputProcessorWrapper(JxlEncoderOutputProcessor processor)
+      : external_output_processor_(
+            jxl::make_unique<JxlEncoderOutputProcessor>(processor)) {}
+
+  bool HasAvailOut() const { return avail_out_ != nullptr; }
+
+  // Caller can never overwrite a previously-written buffer. Asking for a buffer
+  // with `min_size` such that `position + min_size` overlaps with a
+  // previously-written buffer is invalid.
+  jxl::StatusOr<JxlOutputProcessorBuffer> GetBuffer(size_t min_size,
+                                                    size_t requested_size = 0);
+
+  void Seek(size_t pos);
+
+  void SetWatermark(size_t pos);
+
+  size_t CurrentPosition() const { return position_; }
+
+  bool SetAvailOut(uint8_t** next_out, size_t* avail_out);
+
+  bool WasStopRequested() const { return stop_requested_; }
+  bool HasOutputToWrite() const {
+    return output_position_ < watermark_position_;
   }
 
-  {
-    const uint64_t store = large_size ? 1 : box_size;
-    for (size_t i = 0; i < 4; i++) {
-      output->push_back(store >> (8 * (3 - i)) & 0xff);
-    }
+ private:
+  void ReleaseBuffer(size_t bytes_used);
+
+  // Tries to write all the bytes up to the watermark position.
+  void FlushOutput();
+
+  bool AppendBufferToExternalProcessor(void* data, size_t count);
+
+  struct InternalBuffer {
+    // Bytes in the range `[consumed_bytes, written_bytes)` need to be flushed
+    // out.
+    size_t written_bytes = 0;
+    // If data has been buffered, it is stored in `owned_data`.
+    jxl::PaddedBytes owned_data;
+  };
+
+  // Invariant: `internal_buffers_` does not contain chunks that are entirely
+  // below the output position.
+  std::map<size_t, InternalBuffer> internal_buffers_;
+
+  uint8_t** next_out_ = nullptr;
+  size_t* avail_out_ = nullptr;
+  size_t position_ = 0;
+  size_t watermark_position_ = 0;
+  // Either the position of the `external_output_processor_` or the position
+  // `next_out_` points to.
```

Comment:
what is the difference between position_ and output_position_ ?

Reply:
If we support seeking, `output_position_ == position_`.
Otherwise `output_position` is smaller or equal to `position` and points to the beginning of the current buffer, and `position_` where we write next, I think 
@veluca93

Reply:
position_ = where the next GetBuffer call will write bytes to
output_position_ = the position for either the user-given output_processor, or `next_out`.

Reply:
added comments...

Reply:
so now this comment should be updated because output_position_ is not the position of the external output processor, since it is not incremented after a release_buffer call

Reply:
Still not sure if the correct thing is to increase output_position_ by bytes_used